src/axom/klee/Geometry.hpp

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#ifndef AXOM_KLEE_GEOMETRY_HPP
+#define AXOM_KLEE_GEOMETRY_HPP
+
+#include <memory>
+#include <string>
+
+#include "axom/klee/Dimensions.hpp"
+#include "axom/klee/Units.hpp"
+
+namespace axom
+{
+namespace klee
+{
+class GeometryOperator;
+
+/**
+ * Properties of a geometric object which can be transformed by operators
+ */
+struct TransformableGeometryProperties
+{
+  Dimensions dimensions;
+  LengthUnit units;
+};
+
+/**
+ * Compare transformable properties for equality.
+ * \param lhs the left-hand-side operand
+ * \param rhs the right-hand-side operand
+ * \return true if and only if all properties are equal
+ */
+bool operator==(const TransformableGeometryProperties &lhs,
+                const TransformableGeometryProperties &rhs);
+
+/**
+ * Compare transformable properties for inequality.
+ * \param lhs the left-hand-side operand
+ * \param rhs the right-hand-side operand
+ * \return false if and only if all properties are equal
+ */
+inline bool operator!=(const TransformableGeometryProperties &lhs,
+                       const TransformableGeometryProperties &rhs)
+{
+  return !(lhs == rhs);
+}
+
+/**
+ * Represents the geometry specified in a Shape.
+ */
+class Geometry
+{
+public:
+  /**
+   * Create a new Geometry object.
+   *
+   * \param startProperties the transformable properties before any
+   * operators are applied
+   * \param format the format of the file
+   * \param path the path of the file
+   * \param operator_ a possibly null operator to apply to the geometry.
+   */
+  Geometry(const TransformableGeometryProperties &startProperties,
+           std::string format,
+           std::string path,
+           std::shared_ptr<GeometryOperator const> operator_);
```

Comment:
What component logically "owns" this `GeometryOperator`? If it's the `Geometry` object itself, this argument and the corresponding member variable could probably be made an `std::unique_ptr`.

Reply:
There is actual shared ownership here. Sometimes, the Geometry object holds the only copy. Other times, it is part of a named operator and so ownership is shared by the dictionary of named operators and any geometry object that holds a copy.

Reply:
I was starting to go through the design of this to see if we could change it so that only one place really owns the operator, but then I saw the live usages of `getGeometryOperator()` in production codes. We will have to work with those code teams to coordinate any changes.